src/controller.swift

```diff
@@ -153,9 +175,44 @@ private func removeCtrl(char: UInt32) -> UInt32 {
 /// }
 private func repObjectIMK(_ sender: Any?) -> Any? {
   if let sender = sender as? NSMutableDictionary {
-    if let menuItem = sender["IMKCommandMenuItem"] as? NSMenuItem {
+    if let menuItem = sender[kIMKCommandMenuItemName] as? NSMenuItem {
       return menuItem.representedObject
     }
   }
   return nil
 }
+
+struct FcitxAction: Codable {
+  let id: Int
+  let name: String
+  let desc: String
+  let checked: Bool?
+  let children: [FcitxAction]?
+  let separator: Bool?
+
+  // Returns a flattened array of the menu item and all of its children.
+  // Cannot use submenus directly because IMK submenus do not work as expected.
+  func toMenuItems(target: AnyObject, _ depth: Int = 0) -> [NSMenuItem] {
+    if separator ?? false {
+      // Separators should be skipped in a flattened view.
+      return []
+    }
+
+    let item = NSMenuItem(
+      title: String(repeating: "　", count: depth) + (depth > 0 ? "　▶" : "") + desc,
```

Comment:
I personally think this character is not pretty. What do you think of using 2-space indent without this character?

Reply:
Looks good. Will change later.

<img width="189" alt="image" src="https://github.com/fcitx-contrib/fcitx5-macos/assets/23358293/379f64b3-f6b6-42c3-af2c-bff1aa74fed0">